pkg/controller/jobframework/reconciler.go

```diff
@@ -449,15 +451,8 @@ func (r *JobReconciler) ensureOneWorkload(ctx context.Context, job GenericJob, o
 			// than one workload...
 			w = toDelete[0]
 		}
-
-		if _, finished := job.Finished(); finished {
-			if err := r.finalizeJob(ctx, job); err != nil {
-				return nil, fmt.Errorf("finalizing job with no matching workload: %w", err)
-			}
-		} else {
-			if err := r.stopJob(ctx, job, w, StopReasonNoMatchingWorkload, "No matching Workload"); err != nil {
-				return nil, fmt.Errorf("stopping job with no matching workload: %w", err)
-			}
+		if err := r.stopJob(ctx, job, w, StopReasonNoMatchingWorkload, "No matching Workload"); err != nil {
```

Comment:
I think we should only call stop if it's not already finished. Otherwise it's just a waste of API calls.

Is there a case where we need this for pod groups? It doesn't look like it.

Reply:
Hypothetically, a job could be finished but unsuspended. I don't think it's possible for any of the existing integrations, but it could be possible from the `GeneticJob` interface perspective. 

If we're trying to minimize the number of API calls in this case, shouldn't we call `IsSuspended` here instead of `Finished`?

Reply:
So `Finished==true` will always mean that no pods related to the job are active? 

For the current reconciler, workload will be recreated. So if a job is finished, but not suspended, and new workload is not admitted, the job will be stopped:
https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/reconciler.go#L367

With this implementation we have to be sure that the job is stopped when we find out that workload is not found.